language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionRouter.java

```diff
@@ -78,11 +81,15 @@
                 }
             });
         }
-
+        Project project = ctx.workspace().project(ctx.filePath()).orElseThrow();
+        Path projectRoot = (project.kind() == ProjectKind.SINGLE_FILE_PROJECT)
+                ? project.sourceRoot().getParent() :
+                project.sourceRoot();
         // Get available diagnostics based code-actions
         ctx.allDiagnostics().stream().
-                filter(diag -> CommonUtil.isWithinRange(ctx.cursorPosition(),
-                                                        CommonUtil.toRange(diag.location().lineRange())))
+                filter(diag -> projectRoot.resolve(diag.location().lineRange().filePath()).equals(ctx.filePath()) &&
```

Comment:
What if, without doing the filtering at this point, we do the filtering within the context API (`allDiagnostics `)? This is because the code action is associated with a single file. Is there a specific reason where we will have to access all the diagnostics? Otherwise we are going through all the module diagnostics again and again

Reply:
Hmm yeah true, there are other api usage for `ctx.allDiagnostics()`. let's have a separate api for a particular file diagnostics. Shall i send it in another PR?

Reply:
+1